pkg/logql/log/parser.go

```diff
@@ -53,139 +52,110 @@ func (j *JSONParser) Process(line []byte, lbs *LabelsBuilder) ([]byte, bool) {
-	j.buf = append(j.buf, sanitizeLabelKey(field, false)...)
-	// if matches keep going
-	if j.lbs.ParserLabelHints().ShouldExtractPrefix(unsafeGetString(j.buf)) {
-		return string(j.buf), true
-	}
-	return "", false
-}
-
-func (j *JSONParser) parseLabelValue(iter *jsoniter.Iterator, prefix, field string) {
-	// the first time we use the field as label key.
-	if len(prefix) == 0 {
-		key, ok := j.keys.Get(unsafeGetBytes(field), func() (string, bool) {
-			field = sanitizeLabelKey(field, true)
-			if !j.lbs.ParserLabelHints().ShouldExtract(field) {
-				return "", false
+		keys := make([]string, 0)
+		keys = append(keys, field)
+
+		keyArray := make([]string, 0)
+		jsonSpacerIndex := 0
+		for {
+			v, _, _, e := jsonparser.Get(line, keys...)
+			if e != nil {
+				if e != jsonparser.KeyPathNotFoundError {
+					return e
+				}
+				if !strings.ContainsRune(field, jsonSpacer) {
+					break
+				}
+				if jsonSpacerIndex == 0 {
+					keyArray = strings.Split(field, "_")
+				}
+				if jsonSpacerIndex > len(keyArray) {
+					break
+				}
+				keys = keys[0:0]
+				subKey := ""
+				for idx, key := range keyArray {
+					if idx <= jsonSpacerIndex {
+						keys = append(keys, key)
+						continue
+					}
+					if idx > 1 {
+						subKey += "_" + key
+					} else {
+						subKey += key
+					}
+				}
+				keys = append(keys, subKey)
+				jsonSpacerIndex++
+				continue
 			}
-			if j.lbs.BaseHas(field) {
-				field = field + duplicateSuffix
+			err := j.setJSONVal(field, v)
+			if err != nil {
+				return err
 			}
-			return field, true
-		})
-		if !ok {
-			iter.Skip()
-			return
-		}
-		j.lbs.Set(key, readValue(iter))
-		return
-
-	}
-	// otherwise we build the label key using the buffer
-	j.buf = j.buf[:0]
-	j.buf = append(j.buf, prefix...)
-	j.buf = append(j.buf, byte(jsonSpacer))
-	j.buf = append(j.buf, sanitizeLabelKey(field, false)...)
-	key, ok := j.keys.Get(j.buf, func() (string, bool) {
-		if j.lbs.BaseHas(string(j.buf)) {
-			j.buf = append(j.buf, duplicateSuffix...)
-		}
-		if !j.lbs.ParserLabelHints().ShouldExtract(string(j.buf)) {
-			return "", false
+			break
 		}
-		return string(j.buf), true
-	})
-	if !ok {
-		iter.Skip()
-		return
 	}
-	j.lbs.Set(key, readValue(iter))
+	return nil
 }
 
-func (j *JSONParser) RequiredLabelNames() []string { return []string{} }
-
-func readValue(iter *jsoniter.Iterator) string {
-	switch iter.WhatIsNext() {
-	case jsoniter.StringValue:
-		v := iter.ReadString()
-		// the rune error replacement is rejected by Prometheus, so we skip it.
-		if strings.ContainsRune(v, utf8.RuneError) {
-			return ""
-		}
-		return v
-	case jsoniter.NumberValue:
-		return iter.ReadNumber().String()
-	case jsoniter.BoolValue:
-		if iter.ReadBool() {
-			return trueString
-		}
-		return falseString
-	default:
-		iter.Skip()
-		return ""
+func (j *JSONParser) setJSONVal(field string, v []byte) error {
+	if j.lbs.BaseHas(field) {
+		field = field + duplicateSuffix
 	}
+	j.lbs.Set(field, string(v))
```

Comment:
If JSON rune handling was required a replace could be performed here. Either as a string replace, or probably better as a custom string builder

Perhaps something like:
```
var sb strings.Builder
for _, c := range string(v) {
  if c != '\uFFFD' {
    sb.WriteRune(c)
  }
}
return j.lbs.Set(field, sb.String())
```

Reply:
Alternatively to support escaping in general (which currently you don't)

You could try calling `GetString(v)` this should also solve the failing escaping test.

Reply:
thanks, my cassandra keeps crashing this week and I'm busy with two issues this week (https://github.com/grafana/loki/issues/5667 and https://github.com/grafana/loki/issues/5595) , when my cassandra is stable, I'll come back to fix the json parser's with your tips

Reply:
What do you think about an unsafe to string here? This would avoid copying the data until `LabelResult.String()` is called. It's tricky though. The underlying `line` must not change and `LabelResult.String()` would have to unescape. In any case, we must start with some memory profiling.